<a href="https://colab.research.google.com/github/the-goldengate/seattle-airbnb/blob/Rizdky/05_12_24_Stage1_FinalProject_Rakamin(1).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **DATA PREPARATION**

# DATASET DESCRIPTION



# Dataset Listings Description :

**Dataset listings berisi 3.818 sampel. Dataset ini mengandung 92 fitur.**

ID dan Informasi Listing

- id - Identifikasi unik untuk setiap listing.
- listing_url - URL dari listing di Airbnb.
- scrape_id - Identifikasi untuk proses pengambilan data (scraping).
- last_scraped - Tanggal saat data terakhir kali diambil dari Airbnb.
- name - Nama properti yang ditampilkan di platform.
- summary - Deskripsi singkat mengenai listing.
- space - Deskripsi tentang ruang akomodasi.
- description - Deskripsi mendetail mengenai listing.

Informasi Host

- host_id - Identifikasi unik untuk host yang memiliki listing.
- host_url - URL profil dari host di Airbnb.
- host_name - Nama host yang memiliki properti.
- host_since - Tanggal ketika host pertama kali mulai menyewakan propertinya di Airbnb.
- host_location - Lokasi host.
- host_about - Deskripsi tentang host.
- host_response_time - Waktu rata-rata host dalam merespons permintaan dari tamu.
- host_response_rate - Persentase dari permintaan yang dijawab oleh host.
- host_acceptance_rate - Persentase permintaan yang diterima oleh host.
- host_is_superhost - 1 jika host adalah superhost, 0 jika tidak.
- host_thumbnail_url - URL untuk gambar thumbnail host.
- host_picture_url - URL untuk gambar profil host.
- host_neighbourhood - Nama lingkungan dari host.
- host_listings_count - Jumlah listing yang dimiliki oleh host.
- host_total_listings_count - Total jumlah listing yang dimiliki oleh host.

Lokasi dan Jenis Properti

- street - Alamat jalan dari listing.
- neighbourhood - Nama lingkungan dari listing.
- neighbourhood_cleansed - Nama lingkungan yang telah dibersihkan untuk konsistensi.
- neighbourhood_group_cleansed - Nama kelompok lingkungan yang telah dibersihkan.
- city - Kota tempat listing berada.
- state - Negara bagian tempat listing berada.
- zipcode - Kode pos untuk lokasi listing.
- market - Pasar yang terkait dengan listing.
- smart_location - Lokasi yang cerdas, berdasarkan data geo.
- country_code - Kode negara tempat listing berada.
- country - Nama negara tempat listing berada.
- latitude - Koordinat lintang lokasi listing.
- longitude - Koordinat bujur lokasi listing.
- is_location_exact - Indikator apakah lokasi listing tepat (t) atau tidak (f).
- property_type - Jenis properti (misalnya, Apartment, House).
- room_type - Tipe kamar (Entire home/apt, Private Room, Shared Room).

Kapasitas dan Fasilitas

- accommodates - Jumlah tamu maksimum yang dapat ditampung.
- bathrooms - Jumlah kamar mandi yang tersedia.
- bedrooms - Jumlah kamar tidur yang tersedia.
- beds - Jumlah total tempat tidur di listing.
- bed_type - Tipe tempat tidur (misalnya, Real Bed, Futon).
- amenities - Fasilitas yang disediakan (misalnya, TV, Wi-Fi).
- square_feet - Luas properti dalam kaki persegi (jika ada).

Harga dan Ketersediaan

- price - Harga sewa per malam (dalam dolar).
- weekly_price - Harga sewa per minggu (dalam dolar).
- monthly_price - Harga sewa per bulan (dalam dolar).
- security_deposit - Deposit keamanan yang diperlukan (jika ada).
- cleaning_fee - Biaya pembersihan yang dibebankan (jika ada).
- guests_included - Jumlah tamu yang termasuk dalam harga sewa.
- extra_people - Biaya tambahan per tamu tambahan.
- minimum_nights - Jumlah minimum malam untuk pemesanan.
- maximum_nights - Jumlah maksimum malam untuk pemesanan.
- calendar_updated - Tanggal terakhir update kalender.
- has_availability - Indikator apakah listing memiliki ketersediaan (t) atau tidak (f).
- availability_30 - Jumlah hari dalam 30 hari berikutnya ketika listing tersedia.
- availability_60 - Jumlah hari dalam 60 hari berikutnya ketika listing tersedia.
- availability_90 - Jumlah hari dalam 90 hari berikutnya ketika listing tersedia.
- availability_365 - Jumlah hari dalam setahun ketika listing tersedia.

Ulasan dan Rating

- number_of_reviews - Jumlah total ulasan yang diperoleh listing.
- first_review - Tanggal ulasan pertama yang diterima.
- last_review - Tanggal ulasan terakhir yang diterima.
- review_scores_rating - Skor rating dari ulasan.
- review_scores_accuracy - Skor akurasi dari ulasan.
- review_scores_cleanliness - Skor kebersihan dari ulasan.
- review_scores_checkin - Skor check-in dari ulasan.
- review_scores_communication - Skor komunikasi dari ulasan.
- review_scores_location - Skor lokasi dari ulasan.
- review_scores_value - Skor nilai dari ulasan.

Kebijakan dan Verifikasi

- requires_license - Indikator apakah listing memerlukan lisensi (t) atau tidak (f).
- license - Nomor lisensi yang diperlukan (jika ada).
- jurisdiction_names - Nama yurisdiksi di mana listing berada.
- instant_bookable - Indikator apakah listing dapat dipesan secara instan (t) atau tidak (f).
- cancellation_policy - Kebijakan pembatalan untuk listing.
- require_guest_profile_picture - Indikator apakah memerlukan gambar profil tamu (t) atau tidak (f).
- require_guest_phone_verification - Indikator apakah memerlukan verifikasi telepon tamu (t) atau tidak (f).
- calculated_host_listings_count - Jumlah listing yang dimiliki oleh host.
- reviews_per_month - Rata-rata jumlah ulasan per bulan.

# Dataset Reviews Description :

**Dataset reviews berisi 84.849 sampel. Ini mencakup 6 fitur "**

Informasi Ulasan

- listing_id - Identifikasi untuk listing yang direview.
- id - Identifikasi unik untuk ulasan.
- date - Tanggal saat ulasan diajukan.
- reviewer_id - Identifikasi unik untuk reviewer.
- reviewer_name - Nama reviewer.
- comments - Teks umpan balik yang diberikan oleh reviewer.

# Dataset Calendar Description :

**Dataset calendar berisi 1.393.570 sampel. Ini terdiri dari 4 fitur :**

Informasi Ketersediaan dan Harga

- listing_id - Identifikasi untuk listing.
- date - Tanggal untuk ketersediaan.
- available - Status ketersediaan (t untuk tersedia, f untuk tidak tersedia).
- price - Harga untuk tanggal tersebut (dalam dolar, bisa null untuk tanggal yang tidak tersedia).

#Import Libraries

In [ ]:
import gdown
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import folium
import plotly.express as px
from folium import plugins
from folium.plugins import HeatMap
from scipy.stats import pointbiserialr

#Import & Load Dataset

In [ ]:
# Import Dataset
dataset_link = 'https://drive.google.com/drive/folders/1wLAk6IsUNCWhGENAmAUnmXZxLiiv7m4q?usp=sharing'
gdown.download_folder(dataset_link, quiet=True, use_cookies=False)

# Path Dataset
path = '/content/Dataset/'

# Get the data from Dataset Folder
csv_files = [os.path.join(path, file) for file in os.listdir(path) if file.endswith('.csv')]
df = {}

# Load each CSV file into a DataFrame
for file_name in csv_files:
    base_name = os.path.basename(file_name).replace('.csv', '')
    print(f"Load Path: {file_name} into DataFrame: {base_name}..")

    # Load the file and store it
    df[base_name] = pd.read_csv(file_name)

print("\nList Dataset")
print(df.keys())

Load Path: /content/Dataset/calendar.csv into DataFrame: calendar..
Load Path: /content/Dataset/reviews.csv into DataFrame: reviews..
Load Path: /content/Dataset/listings.csv into DataFrame: listings..

List Dataset
dict_keys(['calendar', 'reviews', 'listings'])


In [ ]:
reviews = pd.read_csv('/content/Dataset/reviews.csv')
calendar = pd.read_csv('/content/Dataset/calendar.csv')
listings = pd.read_csv('/content/Dataset/listings.csv')

# **DATA UNDERSTANDING**

##Dataset Reviews

In [ ]:
df['reviews'].head()

listing_id        id        date  reviewer_id reviewer_name  \
0     7202016  38917982  2015-07-19     28943674        Bianca   
1     7202016  39087409  2015-07-20     32440555         Frank   
2     7202016  39820030  2015-07-26     37722850           Ian   
3     7202016  40813543  2015-08-02     33671805        George   
4     7202016  41986501  2015-08-10     34959538          Ming   

                                            comments  
0  Cute and cozy place. Perfect location to every...  
1  Kelly has a great room in a very central locat...  
2  Very spacious apartment, and in a great neighb...  
3  Close to Seattle Center and all it has to offe...  
4  Kelly was a great host and very accommodating ...

In [ ]:
df['reviews'].columns

Index(['listing_id', 'id', 'date', 'reviewer_id', 'reviewer_name', 'comments'], dtype='object')

##Dataset Calendar

In [ ]:
df['calendar'].head()

listing_id        date available   price
0      241032  2016-01-04         t  $85.00
1      241032  2016-01-05         t  $85.00
2      241032  2016-01-06         f     NaN
3      241032  2016-01-07         f     NaN
4      241032  2016-01-08         f     NaN

In [ ]:
df['calendar'].columns

Index(['listing_id', 'date', 'available', 'price'], dtype='object')

##Dataset Listings

In [ ]:
df['listings'].head()

id                           listing_url       scrape_id last_scraped  \
0   241032   https://www.airbnb.com/rooms/241032  20160104002432   2016-01-04   
1   953595   https://www.airbnb.com/rooms/953595  20160104002432   2016-01-04   
2  3308979  https://www.airbnb.com/rooms/3308979  20160104002432   2016-01-04   
3  7421966  https://www.airbnb.com/rooms/7421966  20160104002432   2016-01-04   
4   278830   https://www.airbnb.com/rooms/278830  20160104002432   2016-01-04   

                                  name  \
0         Stylish Queen Anne Apartment   
1   Bright & Airy Queen Anne Apartment   
2  New Modern House-Amazing water view   
3                   Queen Anne Chateau   
4       Charming craftsman 3 bdm house   

                                             summary  \
0                                                NaN   
1  Chemically sensitive? We've removed the irrita...   
2  New modern house built in 2013.  Spectacular s...   
3  A charming apartment that sits atop Queen Anne...   
4  Cozy family craftman house in beautiful neighb...   

                                               space  \
0  Make your self at home in this charming one-be...   
1  Beautiful, hypoallergenic apartment in an extr...   
2  Our house is modern, light and fresh with a wa...   
3                                                NaN   
4  Cozy family craftman house in beautiful neighb...   

                                         description experiences_offered  \
0  Make your self at home in this charming one-be...                none   
1  Chemically sensitive? We've removed the irrita...                none   
2  New modern house built in 2013.  Spectacular s...                none   
3  A charming apartment that sits atop Queen Anne...                none   
4  Cozy family craftman house in beautiful neighb...                none   

                               neighborhood_overview  \
0                                                NaN   
1  Queen Anne is a wonderful, truly functional vi...   
2  Upper Queen Anne is a charming neighborhood fu...   
3                                                NaN   
4  We are in the beautiful neighborhood of Queen ...   

                                               notes  \
0                                                NaN   
1  What's up with the free pillows?  Our home was...   
2  Our house is located just 5 short blocks to To...   
3                                                NaN   
4                                          Belltown    

                                             transit  \
0                                                NaN   
1  Convenient bus stops are just down the block, ...   
2  A bus stop is just 2 blocks away.   Easy bus a...   
3                                                NaN   
4  The nearest public transit bus (D Line) is 2 b...   

                                       thumbnail_url  \
0                                                NaN   
1  https://a0.muscache.com/ac/pictures/14409893/f...   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                          medium_url  \
0                                                NaN   
1  https://a0.muscache.com/im/pictures/14409893/f...   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                         picture_url  \
0  https://a1.muscache.com/ac/pictures/67560560/c...   
1  https://a0.muscache.com/ac/pictures/14409893/f...   
2  https://a2.muscache.com/ac/pictures/b4324e0f-a...   
3  https://a0.muscache.com/ac/pictures/94146944/6...   
4  https://a1.muscache.com/ac/pictures/6120468/b0...   

                                      xl_picture_url   host_id  \
0                                                

In [ ]:
# Mengatur opsi untuk menampilkan semua kolom
pd.set_option('display.max_columns', None)  # Menampilkan semua kolom
display(listings)

id                            listing_url       scrape_id  \
0       241032    https://www.airbnb.com/rooms/241032  20160104002432   
1       953595    https://www.airbnb.com/rooms/953595  20160104002432   
2      3308979   https://www.airbnb.com/rooms/3308979  20160104002432   
3      7421966   https://www.airbnb.com/rooms/7421966  20160104002432   
4       278830    https://www.airbnb.com/rooms/278830  20160104002432   
...        ...                                    ...             ...   
3813   8101950   https://www.airbnb.com/rooms/8101950  20160104002432   
3814   8902327   https://www.airbnb.com/rooms/8902327  20160104002432   
3815  10267360  https://www.airbnb.com/rooms/10267360  20160104002432   
3816   9604740   https://www.airbnb.com/rooms/9604740  20160104002432   
3817  10208623  https://www.airbnb.com/rooms/10208623  20160104002432   

     last_scraped                                 name  \
0      2016-01-04         Stylish Queen Anne Apartment   
1      2016-01-04   Bright & Airy Queen Anne Apartment   
2      2016-01-04  New Modern House-Amazing water view   
3      2016-01-04                   Queen Anne Chateau   
4      2016-01-04       Charming craftsman 3 bdm house   
...           ...                                  ...   
3813   2016-01-04   3BR Mountain View House in Seattle   
3814   2016-01-04    Portage Bay View!-One Bedroom Apt   
3815   2016-01-04    Private apartment view of Lake WA   
3816   2016-01-04    Amazing View with Modern Comfort!   
3817   2016-01-04            Large Lakefront Apartment   

                                                summary  \
0                                                   NaN   
1     Chemically sensitive? We've removed the irrita...   
2     New modern house built in 2013.  Spectacular s...   
3     A charming apartment that sits atop Queen Anne...   
4     Cozy family craftman house in beautiful neighb...   
...                                                 ...   
3813  Our 3BR/2BA house boasts incredible views of t...   
3814  800 square foot 1 bedroom basement apartment w...   
3815  Very comfortable lower unit. Quiet, charming m...   
3816  Cozy studio condo in the heart on Madison Park...   
3817  All hardwood floors, fireplace, 65" TV with Xb...   

                                                  space  \
0     Make your self at home in this charming one-be...   
1     Beautiful, hypoallergenic apartment in an extr...   
2     Our house is modern, light and fresh with a wa...   
3                                                   NaN   
4     Cozy family craftman house in beautiful neighb...   
...                                                 ...   
3813  Our 3BR/2BA house bright, stylish, and wheelch...   
3814  This space has a great view of Portage Bay wit...   
3815                                                NaN   
3816  Fully furnished unit to accommodate most needs...   
3817                                                NaN   

                                            description experiences_offered  \
0     Make your self at home in this charming one-be...                none   
1     Chemically sensitive? We've removed the irrita...                none   
2     New modern house built in 2013.  Spectacular s...                none   
3     A charming apartment that sits atop Queen Anne...                none   
4     Cozy family craftman house in beautiful neighb...                none   
...                                                 ...                 ...   
3813  Our 3BR/2BA house boasts incredible views of t...                none   
3814  800 square foot 1 bedroom basement apartment w...                none   
3815  Very comfortable lower unit. Quiet, charming m...                none   
3816  Cozy studio condo in the heart on Madison Park...                none   
3817  All hardwood floors, fireplace, 65" TV with Xb...                none   

                                  neighborhood_overview  \
0           

In [ ]:
df['listings'].columns

Index(['id', 'listing_url', 'scrape_id', 'last_scraped', 'name', 'summary',
       'space', 'description', 'experiences_offered', 'neighborhood_overview',
       'notes', 'transit', 'thumbnail_url', 'medium_url', 'picture_url',
       'xl_picture_url', 'host_id', 'host_url', 'host_name', 'host_since',
       'host_location', 'host_about', 'host_response_time',
       'host_response_rate', 'host_acceptance_rate', 'host_is_superhost',
       'host_thumbnail_url', 'host_picture_url', 'host_neighbourhood',
       'host_listings_count', 'host_total_listings_count',
       'host_verifications', 'host_has_profile_pic', 'host_identity_verified',
       'street', 'neighbourhood', 'neighbourhood_cleansed',
       'neighbourhood_group_cleansed', 'city', 'state', 'zipcode', 'market',
       'smart_location', 'country_code', 'country', 'latitude', 'longitude',
       'is_location_exact', 'property_type', 'room_type', 'accommodates',
       'bathrooms', 'bedrooms', 'beds', 'bed_type', 'amenities', '

# **1. Descriptive Statistics**

* A. Apakah ada kolom dengan tipe data kurang sesuai, atau nama kolom dan isinya
kurang sesuai?
* B. Apakah ada kolom yang memiliki nilai kosong? Jika ada, apa saja?
* C. Apakah ada kolom yang memiliki nilai summary agak aneh?
(min/mean/median/max/unique/top/freq)


# 1.1. Data Dictionary Reviews

# **A. Apakah ada kolom dengan tipe data kurang sesuai, atau nama kolom dan isinya kurang sesuai?**

In [ ]:
# Memeriksa informasi umum pada reviews
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84849 entries, 0 to 84848
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   listing_id     84849 non-null  int64 
 1   id             84849 non-null  int64 
 2   date           84849 non-null  object
 3   reviewer_id    84849 non-null  int64 
 4   reviewer_name  84849 non-null  object
 5   comments       84831 non-null  object
dtypes: int64(3), object(3)
memory usage: 3.9+ MB


# 1.1.A Transformasi Data

# 1.1.A.1 **Penyesuaian Tipe Data**

List of Column Types:

**Date** (1 Columns) :
- date - Tanggal saat ulasan diajukan (Tipe: datetime64[ns]).

**Categorical** (4 Columns):
- listing_id - Nominal
- id - Nominal
- reviewer_id - Nominal
- reviewer_name - Nominal

**Object** (1 Columns):
- comments - Teks ulasan (dengan 18 nilai kosong ).

In [ ]:
# Mengubah tipe data kolom listing_id, id, dan reviewer_id menjadi object
reviews['listing_id'] = reviews['listing_id'].astype('category')
reviews['id'] = reviews['id'].astype('category')
reviews['reviewer_id'] = reviews['reviewer_id'].astype('category')
reviews['reviewer_name'] = reviews['reviewer_name'].astype('category')

# Mengubah tipe data kolom date menjadi datetime
reviews['date'] = pd.to_datetime(reviews['date'], errors='coerce')

reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84849 entries, 0 to 84848
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   listing_id     84849 non-null  category      
 1   id             84849 non-null  category      
 2   date           84849 non-null  datetime64[ns]
 3   reviewer_id    84849 non-null  category      
 4   reviewer_name  84849 non-null  category      
 5   comments       84831 non-null  object        
dtypes: category(4), datetime64[ns](1), object(1)
memory usage: 8.3+ MB


- Kolom **listing_id , id , reviewer_id dan reviewer_name** yang semuanya bertipe **int64** diubah menjadi **categorical** karena memungkinkan pengelolaan yang lebih efisien dalam memori dan memudahkan analisis yang berfokus pada pengelompokan, terutama ketika berhadapan dengan kategori unik dari listing, ulasan, dan reviewer. Hal ini juga membantu dalam memperjelas bahwa nilai-nilai ini tidak dimaksudkan untuk operasi matematis, melainkan untuk identifikasi kategori.

- Kolom date bertipe **object** diubah menjadi **datetime** untuk analisis tanggal.

- Kolom comments juga bertipe object, tepat karena berisi teks.

# **B. Apakah ada kolom yang memiliki nilai kosong? Jika ada, apa saja?**

# **1.1.B.1 Checking Duplicate Row**

In [ ]:
# Memeriksa baris duplikat
duplicate_rows = reviews.duplicated()

# Menampilkan jumlah baris duplikat
print(f"Jumlah baris duplikat: {duplicate_rows.sum()}")

# Menampilkan baris duplikat
print("Baris duplikat:")
print(reviews[duplicate_rows])

Jumlah baris duplikat: 0
Baris duplikat:
Empty DataFrame
Columns: [listing_id, id, date, reviewer_id, reviewer_name, comments]
Index: []


Tidak Ada Duplicate Row

# **1.1.B.2 Checking Missing Value**

In [ ]:
# Memeriksa jumlah nilai null di setiap kolom
null_values = reviews.isnull().sum()
(null_values)


listing_id        0
id                0
date              0
reviewer_id       0
reviewer_name     0
comments         18
dtype: int64

# **DATA PREPROCESSING**

In [ ]:
# Menghitung total nilai null dan persentase
nvc = pd.DataFrame(reviews.isnull().sum().sort_values(), columns=['Total Null Values'])
nvc['Percentage'] = (nvc['Total Null Values'] / reviews.shape[0]) * 100
nvc["Data Type"] = [reviews[col].dtype for col in reviews.columns]
nvc.sort_values(by=["Total Null Values", "Percentage"], ascending=False, inplace=True)

# Menampilkan hasil
nvc.style.background_gradient(cmap='Blues')

Observations:

Dataset memiliki 6 kolom dan 84,849 baris data.
Terdapat 2 jenis tipe data yaitu: int64 (4 kolom) dan object (2 kolom).

- Kolom comments memiliki 84,831 nilai non-null , dan 18 nilai null / missing values .

Hal yang harus dilakukan untuk Data Pre-Processing adalah:

- Dengan banyaknya data yang dimiliki, kita perlu menangani Missing Values pada kolom comments

Imputation (Frequent Text) : Mengisi nilai yang hilang dengan teks yang paling sering muncul dalam kolom comments.

In [ ]:
# Menghitung teks yang paling sering muncul dalam kolom comments
most_frequent_comment = reviews['comments'].mode()[0]
most_frequent_comment

'The host canceled this reservation the day before arrival. This is an automated posting.'

In [ ]:
# Mengisi nilai yang hilang dengan teks yang paling sering muncul
reviews['comments'] = reviews['comments'].fillna(most_frequent_comment)

In [ ]:
# Menghitung total nilai null dan persentase
nvc = pd.DataFrame(reviews.isnull().sum().sort_values(), columns=['Total Null Values'])
nvc['Percentage'] = (nvc['Total Null Values'] / reviews.shape[0]) * 100
nvc["Data Type"] = [reviews[col].dtype for col in reviews.columns]
nvc.sort_values(by=["Total Null Values", "Percentage"], ascending=False, inplace=True)

# Menampilkan hasil
nvc.style.background_gradient(cmap='Blues')

In [ ]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84849 entries, 0 to 84848
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   listing_id     84849 non-null  category      
 1   id             84849 non-null  category      
 2   date           84849 non-null  datetime64[ns]
 3   reviewer_id    84849 non-null  category      
 4   reviewer_name  84849 non-null  category      
 5   comments       84849 non-null  object        
dtypes: category(4), datetime64[ns](1), object(1)
memory usage: 8.3+ MB


# **C. Apakah ada kolom yang memiliki nilai summary agak aneh? (min/mean/median/max/unique/top/freq)**

# **1.1.C Statistical Summary**

In [ ]:
# Mendapatkan statistik deskriptif
statistics_reviews = reviews.describe(include='all').T  # Transpose untuk tampilan yang lebih baik

# Menampilkan hasil
display(statistics_reviews)

count   unique  \
listing_id     84849.0   3191.0   
id             84849.0  84849.0   
date             84849      NaN   
reviewer_id    84849.0  75730.0   
reviewer_name    84849    14380   
comments         84849    84136   

                                                             top   freq  \
listing_id                                              208356.0  474.0   
id                                                        3721.0    1.0   
date                                                         NaN    NaN   
reviewer_id                                             206203.0   67.0   
reviewer_name                                              David    870   
comments       The host canceled this reservation the day bef...     75   

                                        mean                  min  \
listing_id                               NaN                  NaN   
id                                       NaN                  NaN   
date           2014-12-26 21:05:14.775660288  2009-06-07 00:00:00   
reviewer_id                              NaN                  NaN   
reviewer_name                            NaN                  NaN   
comments                                 NaN                  NaN   

                               25%                  50%                  75%  \
listing_id                     NaN                  NaN                  NaN   
id                             NaN                  NaN                  NaN   
date           2014-08-09 00:00:00  2015-05-16 00:00:00  2015-08-27 00:00:00   
reviewer_id                    NaN                  NaN                  NaN   
reviewer_name                  NaN                  NaN                  NaN   
comments                       NaN                  NaN                  NaN   

                               max  
listing_id                     NaN  
id                             NaN  
date           2016-01-03 00:00:00  
reviewer_id                    NaN  
reviewer_name                  NaN  
comments                       NaN

Hasil Observasi

**listing_id**
- Count: 84,849 (semua nilai non-null).
- Unique: 3,191 (banyak listing yang memiliki ulasan).
- Top: 208356.0 (listing yang paling banyak diulas).
- Freq: 474 (frekuensi tertinggi dari listing).

**id**
- Count: 84,849 (semua nilai non-null).
- Unique: 84,849 (setiap ulasan memiliki ID unik).
- Top: 3721.0 (nomor ID ulasan tertinggi).
- Freq: 1 (setiap ID unik muncul hanya sekali).

**date**
- Count: 84,849 (semua nilai non-null).
- Value: Menunjukkan rentang waktu ulasan yang diajukan, dengan nilai minimum - - dan maksimum menunjukkan waktu ulasan bervariasi antara 2009 hingga 2016.

**reviewer_id**
- Count: 84,849 (semua nilai non-null).
- Unique: 75,730 (banyak reviewer yang berbeda).
- Top: 206203.0 (ID reviewer yang paling banyak menulis ulasan).
- Freq: 67 (frekuensi tertinggi dari ID reviewer).

**reviewer_name**
- Count: 84,849 (semua nilai non-null).
- Unique: 14,380 (banyak nama pengulas yang berbeda).
- Top: David (nama yang paling umum).
- Freq: 870 (jumlah kemunculan nama David).

**comments**
- Count: 84,849 (semua nilai non-null).
- Unique: 84,136 (nilai unik, banyak komentar).
- Top: Teks komentar yang paling umum.
- Freq: 75 (jumlah kemunculan komentar tersebut).

# **Saran Preprocessing Tipe Data**



- listing_id, id, reviewer_id: Pertimbangkan untuk mengubah ketiga kolom ini menjadi categorical jika akan menganalisis berdasarkan kategori dan tidak memerlukan operasi matematis.

- date: Pastikan tipe ini sudah diubah menjadi datetime, jika belum, lakukan ini untuk analisis waktu.

- Nilai Kosong: Tidak ada nilai kosong dalam dataset ini, sehingga tidak diperlukan imputasi atau penghapusan baris.

- Duplikasi (Jika Ada): Periksa apakah ada ulasan ganda dengan memeriksa kolom kombinasi tertentu seperti listing_id, reviewer_id, dan date untuk memastikan tidak ada duplikat yang mungkin mempengaruhi analisis.

- Pemrosesan Komentar: Cleaning Text: Lakukan pembersihan pada kolom comments, seperti menghapus tanda baca atau mengonversi teks ke huruf kecil untuk analisis sentimen nanti. Imputasi atau Pengolahan: Jika ada tujuan untuk menganalisis sentimen, pertimbangkan untuk melakukan tokenisasi dan mengeluarkan fitur dari komentar.

- Analisis Lebih Lanjut: Jika ingin menganalisis asosiasi antara reviewer dan rating, kita dapat melakukan analisis lebih lanjut berdasarkan reviewer_id.


# **1.2 Data Dictionary Calendar**

# **A. Apakah ada kolom dengan tipe data kurang sesuai, atau nama kolom dan isinya kurang sesuai?**


In [ ]:
# Memeriksa informasi umum pada calendar
calendar.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1393570 entries, 0 to 1393569
Data columns (total 4 columns):
 #   Column      Non-Null Count    Dtype 
---  ------      --------------    ----- 
 0   listing_id  1393570 non-null  int64 
 1   date        1393570 non-null  object
 2   available   1393570 non-null  object
 3   price       934542 non-null   object
dtypes: int64(1), object(3)
memory usage: 42.5+ MB


# **1.2.A Transformasi Data**

# **1.2.A.1 Penyesuaian Tipe Data**

List of Column Types:

**Date** (1 Column):

- date - Tanggal untuk ketersediaan (Tipe: datetime64[ns]).

**Categorical** (2 Columns):

- listing_id - Nominal (Identifikasi unik untuk listing).
- vailable - Nominal (Status ketersediaan: t untuk tersedia, f untuk tidak tersedia).

**Object** (1 Column):
- price - Harga untuk tanggal tersebut (dalam dolar, dengan 459,028 nilai kosong ).



In [ ]:
# Mengubah tipe data bagi kolom yang ada
calendar['listing_id'] = calendar['listing_id'].astype('category')  # Ubah listing_id menjadi kategori
calendar['available'] = calendar['available'].astype('category')    # Ubah available menjadi kategori
calendar['price'] = calendar['price'].replace({'\$':'', ',':''}, regex=True).astype(float)  # Menghapus simbol $ dan mengubah tipe data price

# Mengubah tipe data kolom date menjadi datetime
calendar['date'] = pd.to_datetime(calendar['date'], errors='coerce')

# Memeriksa informasi setelah perubahan tipe data
calendar.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1393570 entries, 0 to 1393569
Data columns (total 4 columns):
 #   Column      Non-Null Count    Dtype         
---  ------      --------------    -----         
 0   listing_id  1393570 non-null  category      
 1   date        1393570 non-null  datetime64[ns]
 2   available   1393570 non-null  category      
 3   price       934542 non-null   float64       
dtypes: category(2), datetime64[ns](1), float64(1)
memory usage: 25.4 MB


- Kolom listing_id dan available yang semuanya bertipe object diubah menjadi categorical karena memungkinkan pengelolaan yang lebih efisien dalam memori dan memudahkan analisis yang berfokus pada pengelompokan, terutama ketika berhadapan dengan kategori unik dari listing.

- Kolom date bertipe object diubah menjadi datetime untuk analisis tanggal. Ini memungkinkan penggunaan fungsi-fungsi berbasis waktu yang lebih efektif.
- Kolom price yang sebelumnya bertipe object diubah menjadi float setelah menghapus simbol dolar dan koma. Ini penting untuk analisis numerik yang melibatkan harga dan memungkinkan perhitungan statistik yang tepat.

# **B. Apakah ada kolom yang memiliki nilai kosong? Jika ada, apa saja?**

# **1.2.B.1 Checking Duplicate Row**

In [ ]:
# Memeriksa baris duplikat
duplicate_rows = calendar.duplicated()

# Menampilkan jumlah baris duplikat
print(f"Jumlah baris duplikat: {duplicate_rows.sum()}")

# Menampilkan baris duplikat
if duplicate_rows.sum() > 0:
    print("Baris duplikat:")
    print(calendar[duplicate_rows])
else:
    print("Tidak ada baris duplikat.")

Jumlah baris duplikat: 0
Tidak ada baris duplikat.


**Tidak Ada Duplicate Row**

# **1.2.B.2 Checking Missing Value**

In [ ]:
null_values = calendar.isnull().sum()
print(null_values)

listing_id         0
date               0
available          0
price         459028
dtype: int64


In [ ]:
# Menghitung total nilai null dan persentase
nvc = pd.DataFrame(null_values.sort_values(), columns=['Total Null Values'])
nvc['Percentage'] = (nvc['Total Null Values'] / calendar.shape[0]) * 100
nvc["Data Type"] = [calendar[col].dtype for col in calendar.columns]
nvc.sort_values(by=["Total Null Values", "Percentage"], ascending=False, inplace=True)

# Menampilkan hasil
nvc.style.background_gradient(cmap='Blues')

**Observations:**

- Dataset memiliki 4 kolom dan 1,393,570 baris data. Terdapat 2 jenis tipe data yaitu: category (3 kolom) dan datetime64[ns] (1 kolom).
- Kolom price memiliki 934,542 nilai non-null , dan 459,028 nilai null / missing values. Ini menunjukkan bahwa sekitar 32.94% dari nilai dalam kolom price hilang.

**Hal yang harus dilakukan untuk Data Pre-Processing adalah:**

Dengan jumlah data yang besar, kita perlu menangani Missing Values pada kolom price .

Imputation (Price Estimate) : Mengisi nilai yang hilang dengan estimasi harga berdasarkan rata-rata atau median harga dari listing yang serupa, jika memungkinkan. Alternatif lainnya adalah menggunakan harga dari data lain untuk mengisi nilai yang hilang, memastikan bahwa analisis tetap akurat.

# ** ??? Harus tanyakan Mentor**

Hal yang harus dilakukan untuk Data Pre-Processing adalah:

Dengan jumlah data yang besar, kita perlu menangani Missing Values pada kolom price .

Imputation (Price Estimate) : Mengisi nilai yang hilang dengan estimasi harga berdasarkan rata-rata atau median harga dari listing yang serupa, jika memungkinkan. Alternatif lainnya adalah menggunakan harga dari data lain untuk mengisi nilai yang hilang, memastikan bahwa analisis tetap akurat.

# **C. Apakah ada kolom yang memiliki nilai summary agak aneh? (min/mean/median/max/unique/top/freq)**


# **1.2.C Statistical Summary**

In [ ]:
# Mendapatkan statistik deskriptif
statistics_calendar = calendar.describe(include='all').T  # Transpose untuk tampilan yang lebih baik

# Menampilkan hasil
display(statistics_calendar)

count  unique     top    freq                           mean  \
listing_id  1393570.0  3818.0  3335.0   365.0                            NaN   
date          1393570     NaN     NaN     NaN  2016-07-04 00:00:00.000000256   
available     1393570       2       t  934542                            NaN   
price        934542.0     NaN     NaN     NaN                     137.944859   

                            min                  25%                  50%  \
listing_id                  NaN                  NaN                  NaN   
date        2016-01-04 00:00:00  2016-04-04 00:00:00  2016-07-04 00:00:00   
available                   NaN                  NaN                  NaN   
price                      10.0                 75.0                109.0   

                            75%                  max        std  
listing_id                  NaN                  NaN        NaN  
date        2016-10-03 00:00:00  2017-01-02 00:00:00        NaN  
available                   NaN                  NaN        NaN  
price                     160.0               1650.0  105.06287

**Hasil Observasi**

**listing_id**

- Count : 1,393,570 (semua nilai non-null).
- Unique : 3,818 (banyak listing yang tersedia).
- Top : 3335.0 (listing yang paling banyak muncul dalam kalender).
- Freq : 365 (frekuensi tertinggi dari listing).

**date**

- Count : 1,393,570 (semua nilai non-null).
- Value : Terdapat rentang waktu dari 2016-01-04 hingga 2017-01-02, menunjukkan ketersediaan data sepanjang tahun.

**available**

- Count : 1,393,570 (semua nilai non-null).
- Unique : 2 (hanya dua status: t untuk tersedia dan f untuk tidak tersedia).
- Top : t (tersedia).
- Freq : 934,542 (frekuensi status tersedia).

**price**

- Count : 934,542 (nilai non-null).
- Unique : Banyak nilai unik, namun detail tidak ada.
- Mean : 137.94 (rata-rata harga).
- Min : 10.0 (harga minimum).
- 25% : 75.0 (kuartil pertama).
- 50% : 109.0 (median).
- 75% : 160.0 (kuartil ketiga).
- Max : 1650.0 (harga maksimum, menunjukkan adanya potensi outlier).
- Std : 105.06 (standar deviasi, menunjukkan variasi harga).


# **Saran Preprocessing Tipe Data**

- listing_id :
Pertimbangkan untuk mengubah kolom ini menjadi categorical jika akan dianalisis berbasis kategori.

- date :
Pastikan tipe ini sudah diubah menjadi datetime untuk analisis waktu yang lebih tepat.

- available :
Sudah bertipe category, tetap pertahankan untuk analisis yang memerlukan pengelompokan berdasarkan status ketersediaan.

- price :
Ubah kolom ini menjadi numerik (float) setelah hapus simbol dolar dan tanda baca, untuk analisis statistik.

- Nilai Kosong :
Periksa untuk mengetahui mengapa price memiliki nilai kosong, apakah disebabkan oleh ketersediaan.

- Duplikasi (Jika Ada) :
Periksa apakah ada duplikat dengan memeriksa kombinasi listing_id dan date untuk memastikan tidak ada entri kalender yang berulang.

- Analisis Lebih Lanjut :
Jika perlu, lakukan analisis lebih lanjut untuk menemukan pola dalam ketersediaan dan harga berdasarkan jenis properti dan waktu.

# **1.3 Data Dictionary Listings**

In [ ]:
# Memeriksa informasi umum pada calendar
listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3818 entries, 0 to 3817
Data columns (total 92 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   id                                3818 non-null   int64  
 1   listing_url                       3818 non-null   object 
 2   scrape_id                         3818 non-null   int64  
 3   last_scraped                      3818 non-null   object 
 4   name                              3818 non-null   object 
 5   summary                           3641 non-null   object 
 6   space                             3249 non-null   object 
 7   description                       3818 non-null   object 
 8   experiences_offered               3818 non-null   object 
 9   neighborhood_overview             2786 non-null   object 
 10  notes                             2212 non-null   object 
 11  transit                           2884 non-null   object 
 12  thumbn

- **Kolom id dan host_id** yang semuanya bertipe **int64** diubah menjadi **categorical** karena memungkinkan pengelolaan yang lebih efisien dalam memori dan memudahkan analisis yang berfokus pada pengelompokan, terutama ketika berhadapan dengan
kategori unik dari listing dan host.

- **Kolom last_scraped** yang bertipe **object** diubah menjadi **datetime** untuk analisis berbasis waktu. Ini memungkinkan penggunaan fungsi-fungsi berbasis waktu yang lebih efektif, seperti menghitung rentang waktu antara berbagai tindakan.

- **Kolom price, weekly_price, monthly_price, security_deposit, dan cleaning_fee **yang sebelumnya bertipe **object** diubah menjadi **float** setelah **menghapus simbol dolar dan tanda baca**. Ini penting untuk analisis numerik yang melibatkan harga dan memungkinkan perhitungan statistik yang tepat.

- **Kolom host_is_superhost** yang bertipe **object** perlu diubah menjadi **boolean** untuk memudahkan analisis. Ini membantu dalam memberikan pemahaman yang lebih baik tentang status superhost dan pengaruhnya terhadap kinerja listing.


# **A. Apakah ada kolom dengan tipe data kurang sesuai, atau nama kolom dan isinya kurang sesuai?**

# **1.3.A Transformasi Data**

# **1.3.A.1 Penyesuaian Tipe Data**

# **??? TANYAKAN MENTOR **


List of Column Types:

**Numerical :**

- id - Identifikasi unik untuk setiap listing.
- scrape_id - Identifikasi untuk proses pengambilan data (scraping).
- host_listings_count - Jumlah listing yang dimiliki oleh host.
- host_total_listings_count - Total jumlah listing yang dimiliki oleh host.
- host_listings_count - Jumlah listing yang dimiliki host.
- bathrooms - Jumlah kamar mandi yang tersedia.
- bedrooms - Jumlah kamar tidur yang tersedia.
- beds - Jumlah total tempat tidur di listing.
- square_feet - Luas properti dalam kaki persegi (jika ada).
- price - Harga sewa per malam (dalam dolar).
- weekly_price - Harga sewa per minggu (dalam dolar).
- monthly_price - Harga sewa per bulan (dalam dolar).
- security_deposit - Deposit keamanan yang diperlukan (dalam dolar, jika ada).
- accommodates - Jumlah tamu maksimum yang dapat ditampung.
- guests_included - Jumlah tamu yang termasuk dalam harga sewa.
- extra_people - Biaya tambahan per tamu tambahan.
- minimum_nights - Jumlah minimum malam untuk pemesanan.
- maximum_nights - Jumlah maksimum malam untuk pemesanan.
- availability_30 - Jumlah hari dalam 30 hari berikutnya ketika listing tersedia.
- availability_60 - Jumlah hari dalam 60 hari berikutnya ketika listing tersedia.
- availability_90 - Jumlah hari dalam 90 hari berikutnya ketika listing tersedia.
- availability_365 - Jumlah hari dalam setahun ketika listing tersedia.
- number_of_reviews - Jumlah total ulasan yang diperoleh listing.
- reviews_per_month - Rata-rata jumlah ulasan per bulan.

boolean:

- host_is_superhost - 1 jika host adalah superhost, 0 jika tidak.
- is_location_exact - Indikator apakah lokasi listing tepat (t untuk ya, f untuk tidak).
- requires_license - Indikator apakah listing memerlukan lisensi (t untuk ya, f untuk tidak).
- instant_bookable - Indikator apakah listing dapat dipesan secara instan (t untuk ya, f untuk tidak).
- has_availability - Indikator apakah listing memiliki ketersediaan (t untuk ya, f untuk tidak).
- host_has_profile_pic - Indikator apakah host memiliki foto profil (t untuk ya, f untuk tidak).
- host_identity_verified - Indikator apakah identitas host telah diverifikasi (t untuk ya, f untuk tidak).
- require_guest_profile_picture - Indikator apakah memerlukan gambar profil tamu (t untuk ya, f untuk tidak).
- require_guest_phone_verification - Indikator apakah memerlukan verifikasi telepon tamu (t untuk ya, f untuk tidak).

Date (1 Columns):**

- last_scraped - Tanggal saat data terakhir kali diambil dari Airbnb (Tipe: datetime64[ns]).

**Categorical :**

- listing_id - Nominal (Identifikasi unik untuk listing).
- property_type - Tipe properti (misalnya, Apartment, House).
- room_type - Tipe kamar (misalnya, Seluruh rumah/apt, Kamar pribadi).
- bed_type - Tipe tempat tidur.
- amenities - Fasilitas yang disediakan.
- host_neighbourhood - Nama lingkungan host.
- street - Alamat jalan dari listing.
- neighbourhood - Nama lingkungan dari listing.
- neighbourhood_cleansed - Nama lingkungan yang telah dibersihkan.
- neighbourhood_group_cleansed - Nama kelompok lingkungan yang telah dibersihkan.
- city - Kota tempat listing berada.
- cancellation_policy - Kebijakan pembatalan untuk listing.



In [ ]:
listings = pd.read_csv('/content/Dataset/listings.csv')

In [ ]:
# Mengubah kolom id dan host_id menjadi categorical
listings['id'] = listings['id'].astype('category')
listings['host_id'] = listings['host_id'].astype('category')

# Mengubah kolom last_scraped menjadi datetime
listings['last_scraped'] = pd.to_datetime(listings['last_scraped'], errors='coerce')

# Mengubah kolom price menjadi float setelah membersihkan nilai
listings['price'] = listings['price'].replace({'\$':'', ',':''}, regex=True).astype(float)
listings['weekly_price'] = listings['weekly_price'].replace({'\$':'', ',':''}, regex=True).astype(float)
listings['monthly_price'] = listings['monthly_price'].replace({'\$':'', ',':''}, regex=True).astype(float)
listings['security_deposit'] = listings['security_deposit'].replace({'\$':'', ',':''}, regex=True).astype(float)
listings['cleaning_fee'] = listings['cleaning_fee'].replace({'\$':'', ',':''}, regex=True).astype(float)

# Mengubah kolom host_is_superhost menjadi boolean
listings['host_is_superhost'] = listings['host_is_superhost'].map({'t': True, 'f': False})

In [ ]:
# Memeriksa informasi umum pada listings
listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3818 entries, 0 to 3817
Data columns (total 92 columns):
 #   Column                            Non-Null Count  Dtype         
---  ------                            --------------  -----         
 0   id                                3818 non-null   category      
 1   listing_url                       3818 non-null   object        
 2   scrape_id                         3818 non-null   int64         
 3   last_scraped                      3818 non-null   datetime64[ns]
 4   name                              3818 non-null   object        
 5   summary                           3641 non-null   object        
 6   space                             3249 non-null   object        
 7   description                       3818 non-null   object        
 8   experiences_offered               3818 non-null   object        
 9   neighborhood_overview             2786 non-null   object        
 10  notes                             2212 non-null 

# **B. Apakah ada kolom yang memiliki nilai kosong? Jika ada, apa saja?**

# **1.3.B.1 Checking Duplicate Row**

In [ ]:
# Memeriksa baris duplikat
duplicate_rows = listings.duplicated()

# Menampilkan jumlah baris duplikat
print(f"Jumlah baris duplikat: {duplicate_rows.sum()}")

# Menampilkan baris duplikat
if duplicate_rows.sum() > 0:
    print("Baris duplikat:")
    print(listings[duplicate_rows])
else:
    print("Tidak ada baris duplikat.")

Jumlah baris duplikat: 0
Tidak ada baris duplikat.


**Tidak ada Duplicate Row**

# **1.3.B.2 Checking Missing Value**

In [ ]:
null_values = listings.isnull().sum()
print(null_values)

id                                    0
listing_url                           0
scrape_id                             0
last_scraped                          0
name                                  0
                                   ... 
cancellation_policy                   0
require_guest_profile_picture         0
require_guest_phone_verification      0
calculated_host_listings_count        0
reviews_per_month                   627
Length: 92, dtype: int64


In [ ]:
# Menghitung total nilai null dan persentase
nvc = pd.DataFrame(null_values.sort_values(), columns=['Total Null Values'])
nvc['Percentage'] = (nvc['Total Null Values'] / listings.shape[0]) * 100
nvc["Data Type"] = [listings[col].dtype for col in listings.columns]
nvc.sort_values(by=["Total Null Values", "Percentage"], ascending=False, inplace=True)

# Menampilkan hasil
nvc.style.background_gradient(cmap='Blues')

**Observations:**

Dataset memiliki 92 kolom dan 3,818 baris data. Terdapat 4 jenis tipe data yaitu: int64 (13 kolom), float64 (17 kolom), object (62 kolom), dan category (2 kolom).

- Kolom summary, space, notes, neighborhood_overview memiliki nilai non-null yang kurang dari total baris, yaitu:
summary : 3,641 non-null
space : 3,249 non-null
neighborhood_overview : 2,786 non-null
notes : 2,212 non-null

Hal ini menunjukkan adanya nilai kosong yang perlu ditangani.

- Kolom price, weekly_price, monthly_price, security_deposit, dan cleaning_fee harus diubah dari tipe object ke numerik (float), setelah menghapus simbol dolar dan koma.

**Hal yang harus dilakukan untuk Data Pre-Processing**:

Dengan banyaknya data yang dimiliki, kita perlu menangani Missing Values pada

- kolom summary, space, notes, dan neighborhood_overview:


Imputation (Frequent Text) : Mengisi nilai yang hilang dengan teks yang paling sering muncul dalam kolom tersebut.

- Untuk kolom price, weekly_price, monthly_price, security_deposit, dan cleaning_fee :


Transformasi tipe data : Menghapus simbol dolar dan tanda baca sebelum mengubah tipe menjadi float untuk analisis numerik yang lebih tepat.

# **DATA PREPROCESSING**

In [ ]:
# Menghitung jumlah nilai null di setiap kolom sebelum preprocessing
null_values_before = listings.isnull().sum()

# Menangani Missing Values

# Mengisi nilai kosong pada kolom summary, space, notes, dan neighborhood_overview
for col in ['summary', 'space', 'notes', 'neighborhood_overview']:
    most_frequent_text = listings[col].mode()[0]  # Menghitung teks yang paling sering muncul
    listings[col] = listings[col].fillna(most_frequent_text)  # Imputasi tanpa inplace

# Mengisi nilai kosong pada kolom license
listings['license'] = listings['license'].fillna("Tidak Diperlukan")  # Imputasi untuk license

# Mengisi nilai kosong pada kolom square_feet dengan median
listings['square_feet'] = listings['square_feet'].fillna(listings['square_feet'].median())  # Imputasi dengan median

# Mengisi nilai kosong untuk kolom harga
price_columns = ['price', 'weekly_price', 'monthly_price', 'security_deposit', 'cleaning_fee']
for col in price_columns:
    listings[col] = listings[col].replace({'\$': '', ',': ''}, regex=True).astype(float)  # Menghapus simbol dan mengubah tipe data
    listings[col] = listings[col].fillna(listings[col].median())  # Imputasi dengan median

# Menghitung jumlah nilai null di setiap kolom setelah preprocessing
null_values_after = listings.isnull().sum()

# Membuat DataFrame untuk menampilkan hasil
nvc = pd.DataFrame(null_values_after.sort_values(), columns=['Total Null Values'])
nvc['Percentage'] = (nvc['Total Null Values'] / listings.shape[0]) * 100
nvc["Data Type"] = [listings[col].dtype for col in listings.columns]
nvc.sort_values(by=["Total Null Values", "Percentage"], ascending=False, inplace=True)

# Menampilkan hasil
nvc.style.background_gradient(cmap='Blues')

Dataset reviews tidak memiliki nilai duplikat.

# **C. Apakah ada kolom yang memiliki nilai summary agak aneh? (min/mean/median/max/unique/top/freq)**

# **1.3.C Statistical Summary**

In [ ]:
# Mendapatkan statistik deskriptif
statistics_listings = listings.describe(include='all').T  # Transpose untuk tampilan yang lebih baik

# Menampilkan hasil
display(statistics_listings)

count  unique  \
id                                3818.0  3818.0   
listing_url                         3818    3818   
scrape_id                         3818.0     NaN   
last_scraped                        3818     NaN   
name                                3818    3792   
...                                  ...     ...   
cancellation_policy                 3818       3   
require_guest_profile_picture       3818       2   
require_guest_phone_verification    3818       2   
calculated_host_listings_count    3818.0     NaN   
reviews_per_month                 3191.0     NaN   

                                                                  top  freq  \
id                                                             3335.0   1.0   
listing_url                       https://www.airbnb.com/rooms/241032     1   
scrape_id                                                         NaN   NaN   
last_scraped                                                      NaN   NaN   
name                                           Capitol Hill Apartment     3   
...                                                               ...   ...   
cancellation_policy                                            strict  1417   
require_guest_profile_picture                                       f  3497   
require_guest_phone_verification                                    f  3443   
calculated_host_listings_count                                    NaN   NaN   
reviews_per_month                                                 NaN   NaN   

                                                           mean  \
id                                                          NaN   
listing_url                                                 NaN   
scrape_id                                      20160104002432.0   
last_scraped                      2016-01-04 00:00:00.000000256   
name                                                        NaN   
...                                                         ...   
cancellation_policy                                         NaN   
require_guest_profile_picture                               NaN   
require_guest_phone_verification                            NaN   
calculated_host_listings_count                         2.946307   
reviews_per_month                                      2.078919   

                                                  min                  25%  \
id                                                NaN                  NaN   
listing_url                                       NaN                  NaN   
scrape_id                            20160104002432.0     20160104002432.0   
last_scraped                      2016-01-04 00:00:00  2016-01-04 00:00:00   
name                                              NaN                  NaN   
...                                               ...                  ...   
cancellation_policy                               NaN                  NaN   
require_guest_profile_picture                     NaN                  NaN   
require_guest_phone_verification                  NaN                  NaN   
calculated_host_listings_count                    1.0                  1.0   
reviews_per_month                                0.02                0.695   

                                                  50%                  75%  \
id                                                NaN                  NaN   
listing_url                                       NaN                  NaN   
scrape_id                            20160104002432.0     20160104002432.0   
last_scraped                      2016-01-04 00:00:00  2016-01-04 00:00:00   
name                                              NaN                  NaN   
...                                               ...                  ...   
cancellation_policy                               NaN                  NaN   
require_guest_profile_picture                     NaN                  NaN   
require_guest_pho

# **Hasil Observasi**

**price** :
- Count : 3,818 (semua nilai non-null).
- Mean : 127.98 (harga sewa per malam).
- Min/Max : Menunjukkan rentang harga, dengan nilai maksimum yang jauh lebih tinggi daripada rata-rata. Ini bisa menunjukkan adanya outlier yang perlu diperiksa.
- Observasi : Harga adalah variabel kunci dalam konteks penyesuaian harga untuk kampanye pemasaran. Pengelolaan outlier dan analisis lebih lanjut mengenai distribusi harga akan membantu dalam pengambilan keputusan.

weekly_price :

- Count : 3,818 (semua nilai non-null).
- Mean : 722.87 (harga sewa per minggu).
- Min/Max : Nilai minimum 100.0 dan maksimum 6,300.0, menunjukkan variasi yang besar.
- Observasi : Perlu melakukan pemeriksaan lebih lanjut untuk outlier pada harga mingguan.

monthly_price :

- Count : 3,818 (semua nilai non-null).
- Mean : 2364.23 (harga sewa per bulan).
- Min/Max : Nilai minimum 500.0 dan maksimum 19,500.0. Ini menunjukkan bahwa beberapa listing mungkin tidak sesuai dengan pasar.
- Observasi : Pengecekan outlier diperlukan.

security_deposit :

- Count : 3,818 (semua nilai non-null).
- Mean : 268.38 (deposit keamanan).
- Min/Max : Dengan nilai minimum 95.0 dan maksimum 5,000.0, ada potensi outlier.
- Observasi : Mempertimbangkan nilai deposit yang tidak biasa dalam analisis risiko dan kebijakan.

cleaning_fee :

- Count : 3,818 (semua nilai non-null).
- Mean : 58.55 (biaya pembersihan).
- Min/Max : Nilai maksimum 300.0 menunjukkan adanya variasi biaya pembersihan yang perlu diperiksa lebih lanjut.
- Observasi : Harus diperiksa lebih lanjut untuk kecerobohan dalam biaya pembersihan.

guests_included :
- Count : 3,818 (semua nilai non-null).
- Mean : 1.67 (jumlah tamu yang termasuk dalam harga sewa).
- Min/Max : Nilai minimum 0.0 dan maksimum 15.0, menunjukkan potensi variasi dalam penetapan jumlah tamu.
- Observasi : Pertimbangkan untuk membuat fitur baru yang menyoroti jumlah maksimum tamu yang diizinkan untuk kebijakan harga yang lebih baik.

host_is_superhost :
- Unique : 2 (1 dan 0).
- Observasi : Kolom ini penting untuk menganalisis faktor yang memengaruhi status superhost dan dampaknya terhadap harga dan okupansi.

number_of_reviews :
- Count : 3,818 (semua nilai non-null).
- Mean : 22.22 (rata-rata jumlah ulasan).
- Observasi : Ulasan berhubungan erat dengan kepercayaan dan penetapan harga.

review_scores_rating:
- Count: 3,171 (nilai tidak null).
- Mean: 94.54 (rata-rata rating).
- Observasi: Rating adalah indikator yang baik untuk kualitas rating sebuah listing. Rata-rata rating yang tinggi (94.54) menunjukkan bahwa kebanyakan tamu memberikan umpan balik positif. Ini juga dapat membantu dalam mendorong calon penyewa untuk memilih listing tersebut.

host_response_time:
- Count: 3,295 (nilai non-null).
- Unique: 4 (kecepatan respons host).
- Top: within an hour (kecepatan respons yang paling umum).
- Freq: 1,692 (jumlah host yang merespons dalam waktu satu jam).
- Observasi: Waktu respons host yang cepat (within an hour) dapat meningkatkan kepuasan tamu dan konversi pemesanan. Frekuensi tinggi untuk waktu respons ini menunjukkan bahwa banyak host berusaha memberikan pelayanan yang responsif.

host_response_rate:
- Count: 3,295 (nilai non-null).
- Unique: 45 (persentase respon).
- Top: 100% (persentase respon yang paling umum).
- Freq: 2,371 (jumlah host yang memiliki 100% respons).
- Observasi: Tingkat respons yang tinggi (100%) menunjukkan bahwa host sangat responsif terhadap permintaan penyewa. Ini dapat menarik lebih banyak tamu dan meningkatkan kepuasan pelanggan.

host_acceptance_rate:
- Count: 3,045 (nilai non-null).
- Unique: 2 (nilai persentase).
- Top: 100% (persentase penerimaan yang paling umum).
- Freq: 3,044 (jumlah host yang memiliki 100% penerimaan).
- Observasi: Tingkat penerimaan yang tinggi menandakan bahwa host menerima hampir semua permintaan pemesanan, yang dapat meningkatkan kepercayaan tamu terhadap listing dan host.

accommodates:
- Count: 3,818 (semua nilai non-null).
- Mean: 3.35 (rata-rata jumlah tamu).
- Min: 1.0 (jumlah minimum tamu).
- Max: 16.0 (jumlah maksimum tamu).
- Std: 1.98 (standar deviasi).
Observasi: Rata-rata kapasitas akomodasi yang dapat ditampung adalah 3.35, menunjukan bahwa sebagian besar listing dirancang untuk kelompok kecil hingga sedang. Variasi standar deviasi yang tinggi (1.98) menunjukkan perbedaan kapasitas antara listing.

bathrooms:
- Count: 3,802 (nilai non-null).
- Mean: 1.26 (rata-rata jumlah kamar mandi).
- Min: 0.0 (jumlah minimum kamar mandi).
- Max: 8.0 (jumlah maksimum kamar mandi).
- Std: 0.59 (standar deviasi).
- Observasi: Rata-rata jumlah kamar mandi yang tersedia adalah 1.26, menunjukkan bahwa sebagian besar properti memiliki setidaknya satu kamar mandi. Adanya beberapa listing dengan 0 kamar mandi mungkin memerlukan verifikasi, karena bisa jadi ini adalah kesalahan atau mencerminkan tipe akomodasi tertentu.

bedrooms:
- Count: 3,812 (nilai non-null).
- Mean: 1.31 (rata-rata jumlah kamar tidur).
- Min: 0.0 (jumlah minimum kamar tidur).
- Max: 7.0 (jumlah maksimum kamar tidur).
- Std: 0.88 (standar deviasi).
- Observasi: Rata-rata jumlah kamar tidur adalah 1.31, menunjukkan bahwa banyak listing merupakan akomodasi kecil sampai menengah. Kemiripan dengan jumlah kamar mandi meningkatkan kenyamanan bagi tamu.

beds:
- Count: 3,817 (nilai non-null).
- Mean: 1.74 (rata-rata jumlah tempat tidur).
- Min: 1.0 (jumlah minimum tempat tidur).
- Max: 15.0 (jumlah maksimum tempat tidur).
- Std: 1.14 (standar deviasi).
- Observasi: Jumlah rata-rata tempat tidur adalah 1.74, menunjukkan bahwa sebagian besar listing memiliki minimal satu tempat tidur. Ini sangat relevan untuk menentukan kapasitas akomodasi yang sesuai dengan kebutuhan tamu.

#Univariate Analysis

Gunakan visualisasi untuk melihat distribusi masing-masing kolom (feature maupun
target). Tuliskan hasil observasinya, misalnya jika ada suatu kolom yang distribusinya
menarik (misal skewed, bimodal, ada outlier, ada nilai yang mendominasi, kategorinya
terlalu banyak, dsb). Jelaskan juga apa yang harus di-follow up saat data pre-processing.

# **Distribution Numeric**

# **Distribution price, weekly_price, monthly_price, number_of_reviews**

# **Distribution price**

In [ ]:
# Fungsi untuk membuat histogram dengan boxplot
def histogram_plot(data, column, color):
    fig = px.histogram(
        data,
        x=column,
        marginal='box',
        color_discrete_sequence=[color],
        nbins=75
    )

    fig.update_traces(
        marker_line_width=1,
        marker_line_color='white'
    )

    fig.update_layout(
        plot_bgcolor='rgba(0, 0, 0, 0)',
        title=dict(
            text=f"<b>Distribusi Kolom <span style='color:{color}'>{column.title()}</span></b><br><sup>Listings Settle AirBnB</sup>",
            font=dict(
                size=28,
                color='#757882'
            ),
            y=0.92,
            x=0.5
        ),
        yaxis=dict(
            title='Count',
            showgrid=False,
            showline=False,
            showticklabels=True,
            zeroline=False
        ),
        margin=dict(
            t=80,
            b=10,
            r=20
        )
    )

    fig.show()

    # Tampilkan visualisasi untuk kolom price
histogram_plot(
    data=listings,
    column='price',
    color='#EB8317'  # Warna untuk histogram
)

**Deskripsi**

- Distribusi harga menunjukkan konsentrasi tinggi di bawah 200, mirip dengan pola yang terlihat dalam kolom harga lainnya.

- Outlier menunjukkan beberapa harga tinggi, dengan batas maksimum di sekitar 800.

**Observasi**
- Skewed Distribution : Memiliki distribusi miring ke kanan, menunjukkan banyak listing berada di harga terjangkau, tetapi dengan beberapa harga sangat tinggi sebagai pengecualian.

- Outliers : Terdapat outlier harga yang akan mempengaruhi analisis statistik dan keputusan harga untuk pemilik properti.

**Follow-up**
Pertimbangkan untuk menghapus atau mengelola outlier ini. Periksa faktor-faktor yang menyebabkan variasi harga dan lakukan analisis lebih lanjut untuk memahami pengaturan harga.

# **Distribution Weekly Price**

In [ ]:
# Fungsi untuk membuat histogram dengan boxplot
def histogram_plot(data, column, color):
    fig = px.histogram(
        data,
        x=column,
        marginal='box',
        color_discrete_sequence=[color],
        nbins=75
    )

    fig.update_traces(
        marker_line_width=1,
        marker_line_color='white'
    )

    fig.update_layout(
        plot_bgcolor='rgba(0, 0, 0, 0)',
        title=dict(
            text=f"<b>Distribusi Kolom <span style='color:{color}'>{column.title()}</span></b><br><sup>Listings Settle AirBnB</sup>",
            font=dict(
                size=28,
                color='#757882'
            ),
            y=0.92,
            x=0.5
        ),
        yaxis=dict(
            title='Count',
            showgrid=False,
            showline=False,
            showticklabels=True,
            zeroline=False
        ),
        margin=dict(
            t=80,
            b=10,
            r=20
        )
    )

    fig.show()

    # Tampilkan visualisasi untuk kolom weekly_price
histogram_plot(
    data=listings,
    column='weekly_price',
    color='#F3C623'  # Warna untuk histogram
)


**Deskripsi**

- Distribusi : Distribusi harga mingguan menunjukkan konsentrasi tinggi di bawah 1.000, dengan puncak frekuensi yang jelas pada rentang harga rendah. Banyak listing memiliki harga mingguan yang relatif terjangkau.
- Outlier : Terdapat beberapa outlier yang terletak di sisi kanan histogram, dengan beberapa harga mingguan mencapai atau melebihi 6.000.

**Observasi**
- Skewed Distribution : Distribusi harga mingguan miring ke kanan (right skewed), menunjukkan bahwa mayoritas listing memiliki harga mingguan yang rendah, tetapi ada beberapa listing dengan harga tinggi yang menciptakan ketidakseimbangan.
- Outliers : Terdapat outlier yang jelas, yang dapat mempengaruhi analisis statistik dan model prediksi. Outlier ini mungkin mencerminkan properti
superior atau tempat akomodasi yang lebih premium.

**Follow-up pada Preprocessing**


Identifikasi dan Tangani Outliers :

- Analyzing Outliers : Periksa penyebab dari harga tinggi ini. Apakah berhubungan dengan lokasi, ukuran, atau fasilitas yang ditawarkan?
- Pengelolaan Outliers : Putuskan apakah outlier ini harus dihapus, atau jika masih relevan untuk analisis, gunakan metode seperti winsorization untuk mengurangi dampak mereka.

**Transformasi Data :**

Jika distribusi menunjukkan skewness yang signifikan, pertimbangkan untuk menggunakan transformasi logaritmik untuk mendekati distribusi normal.
Segmentasi :
Pertimbangkan untuk mengelompokkan harga mingguan ke dalam kategori (misalnya, rendah, sedang, tinggi) untuk analisis selanjutnya, sehingga memudahkan
pemahaman tentang pola harga.

**Analisis Lanjutan :**

Selanjutnya, lakukan analisis hubungan antara weekly_price dengan variabel lain seperti number_of_reviews, room_type, dan status superhost untuk memahami faktor-faktor yang mempengaruhi harga.

# **Distribution Monthly Price**

In [ ]:
# Fungsi untuk membuat histogram dengan boxplot
def histogram_plot(data, column, color):
    fig = px.histogram(
        data,
        x=column,
        marginal='box',
        color_discrete_sequence=[color],
        nbins=75
    )

    fig.update_traces(
        marker_line_width=1,
        marker_line_color='white'
    )

    fig.update_layout(
        plot_bgcolor='rgba(0, 0, 0, 0)',
        title=dict(
            text=f"<b>Distribusi Kolom <span style='color:{color}'>{column.title()}</span></b><br><sup>Listings Settle AirBnB</sup>",
            font=dict(
                size=28,
                color='#757882'
            ),
            y=0.92,
            x=0.5
        ),
        yaxis=dict(
            title='Count',
            showgrid=False,
            showline=False,
            showticklabels=True,
            zeroline=False
        ),
        margin=dict(
            t=80,
            b=10,
            r=20
        )
    )

    fig.show()

    # Tampilkan visualisasi untuk kolom monthly_price
histogram_plot(
    data=listings,
    column='monthly_price',
    color='#4A249D'  # Warna untuk histogram
)

**Deskripsi**

Distribusi : Distribusi harga bulanan menunjukkan konsentrasi tinggi pada nilai di bawah 3.000. Sebagian besar listing memiliki harga bulanan di kisaran ini, dengan puncak yang jelas.
Outlier : Ada outlier di sisi kanan yang menunjukkan beberapa harga bulanan yang sangat tinggi, mencapai hingga 20.000.

**Observasi**

- Skewed Distribution : Distribusi harga bulanan miring ke kanan. Hal ini menunjukkan bahwa mayoritas listing berada dalam rentang harga terjangkau, tetapi terdapat sejumlah kecil listing yang memiliki harga bulanan yang sangat tinggi.
- Outliers : Terdapat outlier signifikan yang dapat mempengaruhi analisis statistik. Outlier ini mungkin berasal dari properti premium atau akomodasi yang tidak biasa.
- Dominasi Nilai : Fokus pada harga rendah menunjukkan bahwa sebagian besar konsumen cenderung memilih opsi yang lebih terjangkau.

**Follow-up pada Preprocessing**

Identifikasi dan Tangani Outliers :

- Analisis Outlier : Periksa penyebab harga tinggi ini. Apakah berhubungan dengan lokasi, ukuran properti, atau fasilitas yang ditawarkan?
-[ Pengelolaan Outliers : Putuskan apakah outlier ini perlu dihapus, atau jika tetap relevan, gunakan metode seperti winsorization yang mengurangi dampak harga ekstrem.

**Transformasi Data :**

Jika distribusi menunjukkan skewness yang signifikan, pertimbangkan untuk melakukan transformasi logaritmik pada kolom monthly_price untuk mendekati distribusi normal, yang membantu dalam analisis lebih lanjut.

**Segmentasi Harga :**

Pertimbangkan untuk mengelompokkan data harga bulanan ke dalam kategori (misalnya, rendah, sedang, tinggi) untuk analisis yang lebih mudah dan lebih informatif.

**Analisis Lanjutan :**

Lakukan analisis tambahan tentang bagaimana harga bulanan berkaitan dengan variabel lain, seperti room_type, number_of_reviews, dan status superhost.

# **Distribution Number of Reviews**

In [ ]:
# Fungsi untuk membuat histogram dengan boxplot
def histogram_plot(data, column, color):
    fig = px.histogram(
        data,
        x=column,
        marginal='box',
        color_discrete_sequence=[color],
        nbins=75
    )

    fig.update_traces(
        marker_line_width=1,
        marker_line_color='white'
    )

    fig.update_layout(
        plot_bgcolor='rgba(0, 0, 0, 0)',
        title=dict(
            text=f"<b>Distribusi Kolom <span style='color:{color}'>{column.title()}</span></b><br><sup>Listings Settle AirBnB</sup>",
            font=dict(
                size=28,
                color='#757882'
            ),
            y=0.92,
            x=0.5
        ),
        yaxis=dict(
            title='Count',
            showgrid=False,
            showline=False,
            showticklabels=True,
            zeroline=False
        ),
        margin=dict(
            t=80,
            b=10,
            r=20
        )
    )

    fig.show()

    # Tampilkan visualisasi untuk kolom number_of_reviews
histogram_plot(
    data=listings,
    column='number_of_reviews',
    color='#10375C'  # Warna untuk histogram
)

**Deskripsi**

- Distribusi :Kita melihat bahwa distribusi jumlah ulasan menunjukkan konsentrasi yang sangat tinggi di bawah 10. Banyak listing yang tidak memiliki ulasan atau hanya sedikit, dengan puncak histogram di area tersebut. Untuk beberapa listing, kita juga menemukan beberapa outlier yang memiliki jumlah ulasan hampir mencapai 500.

- Outlier : Terdapat beberapa outlier di sisi kanan yang menunjukkan beberapa listing memiliki jumlah ulasan yang sangat tinggi, mendekati 500.

Observasi

- Skewed Distribution : Kolom ini jelas terlihat miring ke kanan. Ini mengindikasikan bahwa kebanyakan listing sepertinya tidak terlalu sering diulas, mungkin karena kurangnya popularitas atau mungkin saja merupakan listing baru.
- Outliers : Outlier yang terlihat menunjukkan bahwa ada beberapa listing yang sangat populer. Ini bisa jadi menarik untuk kita teliti lebih dalam. Mengapa beberapa listing ini mendapatkan banyak ulasan? Apa yang membuat mereka lebih menarik bagi pengunjung?
- Dominasi Nilai Rendah : Tingginya jumlah listing dengan ulasan rendah menyoroti potensi masalah dalam strategi pemasaran atau daya tarik listing tersebut. Kita perlu memikirkan cara untuk meningkatkan eksposur listing yang kurang dikenal.


Follow-up pada Preprocessing

- Identifikasi dan Analisis Outliers : Kita perlu menyelidiki listing-listing yang memiliki ulasan tinggi. Mungkin ada faktor tertentu seperti lokasi, jenis properti, atau ulasan positif sebelumnya yang mempengaruhi popularitas. Untuk outlier, kita harus memutuskan apa yang harus dilakukan—apakah akan dihapus atau dianalisis lebih lanjut?

- Transformasi Data : Meskipun kita dapat menerapkan transformasi jika diperlukan, kita perlu mempertimbangkan tujuan analisis kita. Apakah kita lebih ingin fokus pada distribusi keseluruhan atau menganalisis performa listing tertentu?

- Segmentasi : Mungkin kita bisa mengelompokan jumlah ulasan ke dalam kategori, seperti 0-10, 0-100, dan sebagainya. Ini akan membantu kita memahami pola dengan lebih baik dan memberikan wawasan yang lebih mendalam.

- Analisis Lanjutan : Penting bagi kita untuk melakukan analisis terkait hubungan antara jumlah ulasan dan faktor lainnya, seperti harga dan status superhost. Ini bisa membuka potensi strategi baru untuk meningkatkan performa listing.

# **Distribution Categorical**

In [ ]:
# Fungsi untuk membuat histogram dengan boxplot (peruntukan kolom kategorikal)
def histogram_plot(data, column, color):
    fig = px.histogram(
        data,
        x=column,
        color_discrete_sequence=[color],
        title=f'Distribusi Kolom <span style="color:{color}">{column.title()}</span>',
        text_auto=True,
        category_orders=None
    )

    fig.update_layout(
        plot_bgcolor='rgba(0, 0, 0, 0)',
        title=dict(
            text=f"<b>Distribusi Kolom <span style='color:{color}'>{column.title()}</span></b><br><sup>Listings Settle AirBnB</sup>",
            font=dict(
                size=28,
                color='#757882'
            ),
            y=0.92,
            x=0.5
        ),
        yaxis=dict(
            title='Count',
            showgrid=False,
            showline=False,
            showticklabels=True,
            zeroline=False
        ),
        margin=dict(
            t=80,
            b=10,
            r=20
        )
    )

    fig.show()

In [ ]:

# Kolom kategorikal yang ingin ditampilkan dan warna untuk setiap kolom
categorical_columns = ['room_type']
color_plot = ['#F3C623']  # Warna untuk histogram

# Tampilkan visualisasi untuk setiap kolom kategorikal satu per satu
for col, color in zip(categorical_columns, color_plot):
    histogram_plot(
        data=listings,
        column=col,
        color=color
    )

# **Distribution Room Type**

**Deskripsi**

- Distribusi : Distribusi menunjukkan bahwa "Entire home/apt" adalah kategori yang paling banyak, dengan total 2.541 listing. "Private room" mengikuti dengan 1.160, sementara "Shared room" hanya terdiri dari 117 listing. Ini mengindikasikan preferensi yang sangat jelas di antara pengguna untuk akomodasi yang lebih privat.

**Observasi**

- Dominasi Kategori : Tipe "Entire home/apt" mendominasi distribusi, yang menyebabkan ketidakseimbangan dalam dataset. Hal ini menunjukkan bahwa sebagian besar penyewa memprioritaskan privasi dan kenyamanan yang ditawarkan oleh jenis akomodasi ini.
- Ketidakseimbangan : Adanya dominasi yang jelas dari satu kategori dapat mempengaruhi analisis kita, sehingga mempersempit fokus pada hanya satu jenis akomodasi.

**Rekomendasi untuk Follow-Up pada Preprocessing**

- Pengelompokan Kategori :
Untuk penyederhanaan analisis, kita dapat mempertimbangkan untuk mengelompokkan "Private room" dan "Shared room" ke dalam kategori baru bernama "Shared Accommodations." Ini akan mempermudah analisis dan memberikan pandangan yang lebih holistik mengenai preferensi penyewa.
- Analisis Hubungan :
Di tahap berikutnya, penting untuk mengeksplorasi hubungan antara room_type dan variabel lain seperti price, tingkat hunian, dan jumlah ulasan. Hal ini akan membantu dalam memahami faktor-faktor yang mempengaruhi daya tarik masing-masing kategori.
- Strategi Pemasaran :
Mengidentifikasi bahwa "Entire home/apt" lebih populer dapat membantu tim pemasaran dalam menargetkan kampanye mereka dengan lebih efektif pada kategori ini, sambil meningkatkan fokus pada strategi untuk menarik tamu ke "Private room" dan "Shared room."
- Evaluasi Kategori yang Jarang :
Listing dalam kategori yang jarang seperti "Shared room" perlu dievaluasi relevansinya. Jika tampilannya minimal, kita perlu mengevaluasi kebijakan dan strategi yang dapat mempengaruhi daya tarik kategori ini.

In [ ]:
categorical_columns = ['property_type']
color_plot = ['#4A249D']  # Warna untuk histogram

# Tampilkan visualisasi untuk setiap kolom kategorikal satu per satu
for col, color in zip(categorical_columns, color_plot):
    histogram_plot(
        data=listings,
        column=col,
        color=color
    )

# **Distribution Property Type**

Deskripsi

- Distribusi : Dalam visualisasi, kita melihat bahwa ada dua jenis properti yang paling banyak: "House" dengan 1.733 listing dan "Apartment" dengan 1.708 listing. Kategori lainnya seperti "Cabin" (21), "Condominium" (91), dan "Townhouse" (118) memiliki jumlah yang signifikan lebih rendah dibandingkan dengan dua kategori utama tersebut.

Observasi
- Dominasi Nilai : Terlihat jelas bahwa "House" dan "Apartment" mendominasi pasar, menunjukkan preferensi yang kuat di antara penyewa. Ini mungkin mencerminkan tipe akomodasi yang lebih disukai karena ruang dan kenyamanan yang ditawarkan.
- Ketidakseimbangan Kategori : Ada ketidakseimbangan yang mencolok di antara kategori. Misalnya, kategori seperti "Tent" dan "Yurt" hanya memiliki masing-masing 5 dan 1 listing. Ketidakseimbangan ini bisa menyulitkan analisis yang lebih dalam jika terlalu banyak kategori yang memiliki sangat sedikit data.

Follow-up pada Preprocessing
- Pengelompokan Kategori :
Mengingat adanya kategori dengan jumlah listing yang sangat sedikit, kita bisa mempertimbangkan untuk mengelompokkan tipe properti yang kurang umum menjadi kategori yang lebih besar, misalnya mengelompokkan "Camper/RV", "Tent", dan "Yurt" menjadi satu kategori "Lainnya". Ini akan memperjelas analisis dan membuatnya lebih informatif.

- Analisis Keterkaitan :
Penting untuk melihat bagaimana jenis properti berhubungan dengan faktor lain, seperti harga dan rating. Kita harus menganalisis apakah jenis properti tertentu memiliki harga yang lebih tinggi atau lebih rendah, serta perbandingan ulasan yang diterima.
- Strategi Pemasaran :
Dengan mengetahui bahwa "House" dan "Apartment" adalah yang paling diminati, tim pemasaran dapat memfokuskan upaya pada kategori ini. Namun, kita juga perlu mengeksplorasi cara untuk menarik lebih banyak perhatian kepada kategori yang kurang populer.
- Evaluasi Kategori yang Jarang :
Listing dalam kategori jarang seperti "Boat" dan "Dorm" mungkin perlu dievaluasi untuk mencari tahu apakah strategi pemasaran atau penyajian data perlu diubah untuk meningkatkan visibilitas mereka.

In [ ]:
categorical_columns = ['host_is_superhost']
color_plot = ['#EB8317']  # Warna untuk histogram

# Tampilkan visualisasi untuk setiap kolom kategorikal satu per satu
for col, color in zip(categorical_columns, color_plot):
    histogram_plot(
        data=listings,
        column=col,
        color=color
    )

# **Distribution Host is Superhost**

Deskripsi

- Distribusi : Dari visualisasi, tampak jelas bahwa mayoritas listing (3.038) tidak memiliki status superhost (False), sedangkan hanya 778 listing yang memiliki status tersebut (True). Ini menunjukkan bahwa sebagian besar pemilik properti belum memenuhi kriteria untuk mendapatkan status superhost.
Observasi

- Dominasi Nilai : Ada dominasi yang signifikan antara kategori "False" dan "True". Banyak listing yang tidak menjadi superhost, menunjukkan bahwa hanya segelintir host yang berhasil memenuhi standar yang ditetapkan oleh Airbnb.
Ketidakseimbangan Kategori : Ketidakseimbangan ini berpotensi mempengaruhi analisis kita. Banyak host yang tidak mendapatkan status superhost mungkin tidak menyadari manfaat dari status tersebut atau belum tahu langkah-langkah yang diperlukan untuk mencapainya.

Follow-up pada Preprocessing

- Analisis Keterkaitan :
Penting untuk mengeksplorasi hubungan antara status superhost dengan variabel lain seperti price, number_of_reviews, dan room_type. Pengetahuan ini akan membantu kita memahami faktor-faktor yang mendorong pemilik listing untuk mencapai status superhost.

- Edukasi untuk Host :
Mengingat bahwa hanya sedikit host yang menjadi superhost, kita perlu merumuskan strategi edukasi untuk membantu mereka memahami kriteria dan cara untuk memenuhi syarat tersebut. Ini bisa meliputi panduan, webinar, atau dukungan langsung.

- Pemantauan Kinerja :
Lakukan analisis berkelanjutan untuk memantau bagaimana status superhost memengaruhi tingkat hunian dan umpan balik dari tamu. Data ini akan memberikan wawasan tambahan bagi tim pemasaran untuk memodifikasi pendekatan mereka.

- Keputusan Strategis :
Evaluasi bagaimana status superhost berdampak pada keputusan harga dan promosi. Jika banyak host yang tidak menjadi superhost, kita mungkin perlu menyesuaikan kebijakan dukungan untuk menarik lebih banyak host untuk meningkatkan kualitas properti mereka.

# **3. Multivariate Analysis**

Lakukan multivariate analysis (seperti correlation heatmap dan category plots, sesuai yang
diajarkan di kelas). Tuliskan hasil observasinya, seperti:
* Bagaimana korelasi antara masing-masing feature dan label. Kira-kira feature mana
saja yang paling relevan dan harus dipertahankan?
* Bagaimana korelasi antar-feature, apakah ada pola yang menarik? Apa yang perlu
dilakukan terhadap feature itu?

# **Merge Dataset**

In [ ]:
# Agregasi data reviews untuk menghitung jumlah ulasan per listing
reviews_agg = reviews.groupby('listing_id').agg({
    'id': 'count'  # Menghitung total ulasan per listing
}).rename(columns={'id': 'number_of_reviews'}).reset_index()
reviews_agg

<ipython-input-140-549b05314002>:2: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



listing_id  number_of_reviews
0          4291                 35
1          5682                297
2          6606                 52
3          7369                 40
4          9419                 79
...         ...                ...
3186    9995551                  1
3187   10012724                  2
3188   10020221                  1
3189   10118341                  1
3190   10248139                  1

[3191 rows x 2 columns]

In [ ]:
# Verifikasi kolom
print("Columns in Listings:", listings.columns)
print("Columns in Reviews Aggregated:", reviews_agg.columns)

# Jika listings memiliki kolom 'id' dan bukan 'listing_id', ganti di penggabungan
merged_data = pd.merge(listings, reviews_agg, left_on='id', right_on='listing_id', how='left')

merged_data

Columns in Listings: Index(['id', 'listing_url', 'scrape_id', 'last_scraped', 'name', 'summary',
       'space', 'description', 'experiences_offered', 'neighborhood_overview',
       'notes', 'transit', 'thumbnail_url', 'medium_url', 'picture_url',
       'xl_picture_url', 'host_id', 'host_url', 'host_name', 'host_since',
       'host_location', 'host_about', 'host_response_time',
       'host_response_rate', 'host_acceptance_rate', 'host_is_superhost',
       'host_thumbnail_url', 'host_picture_url', 'host_neighbourhood',
       'host_listings_count', 'host_total_listings_count',
       'host_verifications', 'host_has_profile_pic', 'host_identity_verified',
       'street', 'neighbourhood', 'neighbourhood_cleansed',
       'neighbourhood_group_cleansed', 'city', 'state', 'zipcode', 'market',
       'smart_location', 'country_code', 'country', 'latitude', 'longitude',
       'is_location_exact', 'property_type', 'room_type', 'accommodates',
       'bathrooms', 'bedrooms', 'beds', 'bed_

id                            listing_url       scrape_id  \
0       241032    https://www.airbnb.com/rooms/241032  20160104002432   
1       953595    https://www.airbnb.com/rooms/953595  20160104002432   
2      3308979   https://www.airbnb.com/rooms/3308979  20160104002432   
3      7421966   https://www.airbnb.com/rooms/7421966  20160104002432   
4       278830    https://www.airbnb.com/rooms/278830  20160104002432   
...        ...                                    ...             ...   
3813   8101950   https://www.airbnb.com/rooms/8101950  20160104002432   
3814   8902327   https://www.airbnb.com/rooms/8902327  20160104002432   
3815  10267360  https://www.airbnb.com/rooms/10267360  20160104002432   
3816   9604740   https://www.airbnb.com/rooms/9604740  20160104002432   
3817  10208623  https://www.airbnb.com/rooms/10208623  20160104002432   

     last_scraped                                 name  \
0      2016-01-04         Stylish Queen Anne Apartment   
1      2016-01-04   Bright & Airy Queen Anne Apartment   
2      2016-01-04  New Modern House-Amazing water view   
3      2016-01-04                   Queen Anne Chateau   
4      2016-01-04       Charming craftsman 3 bdm house   
...           ...                                  ...   
3813   2016-01-04   3BR Mountain View House in Seattle   
3814   2016-01-04    Portage Bay View!-One Bedroom Apt   
3815   2016-01-04    Private apartment view of Lake WA   
3816   2016-01-04    Amazing View with Modern Comfort!   
3817   2016-01-04            Large Lakefront Apartment   

                                                summary  \
0     This is a modern fully-furnished studio apartm...   
1     Chemically sensitive? We've removed the irrita...   
2     New modern house built in 2013.  Spectacular s...   
3     A charming apartment that sits atop Queen Anne...   
4     Cozy family craftman house in beautiful neighb...   
...                                                 ...   
3813  Our 3BR/2BA house boasts incredible views of t...   
3814  800 square foot 1 bedroom basement apartment w...   
3815  Very comfortable lower unit. Quiet, charming m...   
3816  Cozy studio condo in the heart on Madison Park...   
3817  All hardwood floors, fireplace, 65" TV with Xb...   

                                                  space  \
0     Make your self at home in this charming one-be...   
1     Beautiful, hypoallergenic apartment in an extr...   
2     Our house is modern, light and fresh with a wa...   
3     *Note: This fall, there will be major renovati...   
4     Cozy family craftman house in beautiful neighb...   
...                                                 ...   
3813  Our 3BR/2BA house bright, stylish, and wheelch...   
3814  This space has a great view of Portage Bay wit...   
3815  *Note: This fall, there will be major renovati...   
3816  Fully furnished unit to accommodate most needs...   
3817  *Note: This fall, there will be major renovati...   

                                            description experiences_offered  \
0     Make your self at home in this charming one-be...                none   
1     Chemically sensitive? We've removed the irrita...                none   
2     New modern house built in 2013.  Spectacular s...                none   
3     A charming apartment that sits atop Queen Anne...                none   
4     Cozy family craftman house in beautiful neighb...                none   
...                                                 ...                 ...   
3813  Our 3BR/2BA house boasts incredible views of t...                none   
3814  800 square foot 1 bedroom basement apartment w...                none   
3815  Very comfortable lower unit. Quiet, charming m...                none   
3816  Cozy studio condo in the heart on Madison Park...                none   
3817  All hardwood floors, fireplace, 65" TV with Xb...                none   

                                  neighborhood_overview  \
0     Wallin

In [ ]:
# Ambil harga terbaru dari calendar
calendar['date'] = pd.to_datetime(calendar['date'])
latest_calendar = calendar.loc[calendar.groupby('listing_id')['date'].idxmax()]
latest_calendar

<ipython-input-142-2f6bf3f19e18>:3: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



listing_id       date available  price
797524        3335 2017-01-02         t  120.0
1138434       4291 2017-01-02         t   82.0
790954        5682 2017-01-02         f    NaN
137969        6606 2017-01-02         t   90.0
1006669       7369 2017-01-02         f    NaN
...            ...        ...       ...    ...
1134054   10332096 2017-01-02         t   40.0
1056309   10334184 2017-01-02         t  120.0
1078939   10339144 2017-01-02         f    NaN
635464    10339145 2017-01-02         t  219.0
744234    10340165 2017-01-02         t   43.0

[3818 rows x 4 columns]

In [ ]:
# Gabungkan data harga ke merged_data
merged_data = pd.merge(merged_data, latest_calendar[['listing_id', 'price']], on='listing_id', how='left')

# Verifikasi penggabungan
merged_data

ValueError: Cannot convert float NaN to integer

# **Analisis Korelasi**

##Korelasi Antara Feature & Target

##Korelasi Antara Feature

#Data Cleansing (Lakukan analisis terlebih dahulu)

##Dataset Reviews

In [ ]:
# Menghapus kolom kosong pada reviews
reviews = reviews.dropna(axis = 0, subset = ['comments'], how = 'any')

In [ ]:
# Memeriksa data kosong pada reviews setelah dihapus
reviews.isnull().sum()

* Dataset reviews hanya memiliki data kosong dan tidak memiliki data yang duplikat, jadi cukup hanya dengan menghapus data yang kosong.
* Tidak banyak yang perlu diubah dari kolom reviews karena sudah sesuai.

##Dataset Calendar

In [ ]:
# Menghapus kolom kosong pada calendar
calendar = calendar.dropna(axis = 0, subset = ['price'], how = 'any')

In [ ]:
# Memeriksa data kosong pada calendar setelah dihapus
calendar.isnull().sum()

In [ ]:
# Mengubah data type calendar menjadi datetime
calendar['date'] = pd.to_datetime(calendar['date'])

# Memisahkan kolom date menjadi day, month, dan year
calendar['day'] = calendar.date.dt.day
calendar['month'] = calendar.date.dt.month
calendar['year'] = calendar.date.dt.year

# Menghapus lambang dolar pada price
calendar['price'] = pd.to_numeric(calendar['price'].apply(lambda x: str(x).replace('$', '').replace(',', '')), errors='coerce')

# Mengubah available menjadi tipe data boolean
calendar['available'] = calendar['available'].map({'t': True, 'f': False})

In [ ]:
# Melihat dataset calendar setelah diproses
calendar.head()

* Dataset calendar memiliki data kosong pada kolom price dan banyak sekali baris yang kosong, sehingga perlu dihapus untuk kelancaran analisis.
* Kolom price memiliki simbol "$" dan koma (",") sehingga tidak cocok sebagai object. Lalu konversi kolom price ke tipe numerik setelah membersihkan simbol tersebut.
* Kolom date sebaiknya memiliki tipe data datetime daripada object. Setelah diubah pisahkan day, month, dan year untuk mendapatkan insight yang lebih luas.
* Kolom available masih berisi nilai t/f (true/false) dalam bentuk string, dikonversi ke tipe boolean.

##Dataset Listings

In [ ]:
# Mengganti dan membersihkan datatype yang kurang sesuai
listings['price'] = pd.to_numeric(listings['price'].apply(lambda x: str(x).replace('$', '').replace(',', '')), errors='coerce')
listings['monthly_price'] = pd.to_numeric(listings['monthly_price'].apply(lambda x: str(x).replace('$', '').replace(',', '')), errors='coerce')
listings['weekly_price'] = pd.to_numeric(listings['weekly_price'].apply(lambda x: str(x).replace('$', '').replace(',', '')), errors='coerce')
listings['cleaning_fee'] = pd.to_numeric(listings['cleaning_fee'].apply(lambda x: str(x).replace('$', '').replace(',', '')), errors='coerce')
listings['security_deposit'] = pd.to_numeric(listings['security_deposit'].apply(lambda x: str(x).replace('$', '').replace(',', '')), errors='coerce')
listings['extra_people'] = pd.to_numeric(listings['extra_people'].apply(lambda x: str(x).replace('$', '').replace(',', '')), errors='coerce')

Menghapus lambang dolar pada kolom yang memiliki lambang dolar dan mengubahnya menjadi numerik untuk memudahkan analisis dan visualisasi

In [ ]:
# Mengisi nilai kosong untuk kolom kategori
listings['summary'] = listings['summary'].fillna("Tidak tersedia")
listings['space'] = listings['space'].fillna("Tidak tersedia")
listings['neighborhood_overview'] = listings['neighborhood_overview'].fillna("Tidak tersedia")
listings['notes'] = listings['notes'].fillna("Tidak tersedia")
listings['transit'] = listings['transit'].fillna("Tidak tersedia")

Mengisi kolom kategorik yang kosong menjadi "Tidak tersedia"

In [ ]:
# Mengisi nilai kosong untuk kolom numerik
listings['bathrooms'] = listings['bathrooms'].fillna(listings['bathrooms'].median())
listings['bedrooms'] = listings['bedrooms'].fillna(listings['bedrooms'].median())
listings['review_scores_rating'] = listings['review_scores_rating'].fillna(listings['review_scores_rating'].median())
listings['reviews_per_month'] = listings['reviews_per_month'].fillna(listings['reviews_per_month'].mean())
listings['square_feet'] = listings['square_feet'].fillna(0)
listings['security_deposit'] = listings['security_deposit'].fillna(0)
listings['cleaning_fee'] = listings['cleaning_fee'].fillna(0)

Mengisi kolom numerik
* median digunakan untuk kolom yang mempunyai nilai outlier
* mean digunakan untuk kolom yang memiliki distribusi tidak terlalu miring
* 0 digunakan untuk kolom yang masuk akal diisi dengan data kosong/nol

In [ ]:
# Mengisi kolom URL dengan default
listings[['thumbnail_url', 'medium_url', 'xl_picture_url']] = listings[['thumbnail_url', 'medium_url', 'xl_picture_url']].fillna("No Image")

Jika nilai kosong, isi dengan string "No Image"


In [ ]:
# Mengubah typedata rate menjadi numerik
listings['host_response_rate'] = listings['host_response_rate'].str.rstrip('%').astype(float)
listings['host_acceptance_rate'] = listings['host_acceptance_rate'].str.rstrip('%').astype(float)

Karena ada simbol % sebaiknya diubah menjadi numerik, agar mendapat pengolahan lebih lanjut

In [ ]:
# Menghapus kolom yang terlalu banyak memiliki nilai kosong
listings = listings.drop(columns=['license'])

Masih banyak kolom yang memiliki nilai kosong jadi sebaiknya hanya mengambil beberapa kolom saja dari listings yang sesuai dengan goal dan objective yang ingin dicapai.